python/ray/serve/tests/test_grpc.py

```diff
@@ -218,6 +314,418 @@ def __init__(self):
+    with pytest.raises(grpc.RpcError) as exception_info:
+        _, _ = stub.__call__.with_call(request=request)
+    rpc_error = exception_info.value
+    assert rpc_error.code() == grpc.StatusCode.NOT_FOUND
+    assert "Application '' not found." in rpc_error.details()
+
+
+def test_grpc_proxy_with_request_id(ray_cluster):
+    """Test gRPC request with and without request id.
+
+    When no request id is passed, gRPC proxy will respond with a random request id in
+    the trailing metadata. When request id is passed, gRPC proxy will respond with the
+    original request id.
+    """
+    cluster = ray_cluster
+    cluster.add_node(num_cpus=2)
+    cluster.connect(namespace=SERVE_NAMESPACE)
+
+    grpc_port = 9000
+    grpc_servicer_functions = [
+        "ray.serve.generated.serve_pb2_grpc.add_UserDefinedServiceServicer_to_server",
+        "ray.serve.generated.serve_pb2_grpc.add_FruitServiceServicer_to_server",
+    ]
+
+    serve.start(
+        grpc_options={
+            "port": grpc_port,
+            "grpc_servicer_functions": grpc_servicer_functions,
+        },
+    )
+
+    app1 = "app1"
+    serve.run(target=g, name=app1, route_prefix=f"/{app1}")
+
+    # Ensures the app is not yet deployed.
+    deployment_name = "grpc-deployment"
+    app1_replica_name = f"{app1}_{deployment_name}"
+    replicas = ray.get(
+        serve.context._global_client._controller._all_running_replicas.remote()
+    )
+
+    # Ensures the app is deployed.
+    assert len(replicas[app1_replica_name]) == 1
+
+    channel = grpc.insecure_channel("localhost:9000")
+    stub = serve_pb2_grpc.UserDefinedServiceStub(channel)
+    request = serve_pb2.UserDefinedMessage(name="foo", num=30, foo="bar")
+
+    # Ensures the gRPC Proxy returning the original request id.
+    custom_request_id = "fake-request-id"
+    metadata = (("request_id", custom_request_id),)
+    response, call = stub.__call__.with_call(request=request, metadata=metadata)
+    response_request_id = None
+    for key, value in call.trailing_metadata():
+        if key == "request_id":
+            response_request_id = value
+            break
+    assert custom_request_id == response_request_id
+
+    # Ensures the gRPC Proxy returning a new request id.
+    response, call = stub.__call__.with_call(request=request)
+    response_request_id = None
+    for key, value in call.trailing_metadata():
+        if key == "request_id":
+            response_request_id = value
+            break
+    assert custom_request_id != response_request_id
+
+
+def test_grpc_proxy_on_draining_nodes(ray_cluster):
+    """Test gRPC request on the draining node.
+
+    When there are no replicas on head node and some replicas on the worker node, the
+    ListApplications and Healthz methods should respond successfully. When there are
+    no replicas on any nodes, ListApplications and Healthz methods should continue to
+    succeeding on the head node. But should return draining response on the worker node.
```

Comment:
Why should they succeed on the Head node? What's the idea here?

The problem with that behavior is following: load balancers like AWS ELB are failing open when all of its targets are _unhealthy_ (ie failing health-checks). In this case though, all worker nodes will be failing the health-checks, but head-node won't be meaning that all traffic will be routed from LB to the head-node ultimately crushing it (provided that it also runs cluster's control plane).

Reply:
This is an existing behavior and maybe we can talk offline about it. But this was originally to help fixing the downscaling issue. A node's proxy setting in the draining state should not be getting new requests (thus failing the health check). And once all ongoing requests on this node are completed, it will be safe to take down. We will never want to take down head node, so that's why head node will always be reachable and never be draining. 

PS. just in case you need more context, this only happens when there are NO replicas on the node.

Reply:
Understood. Let's add this context to the description to make it more apparent.